feat : 듀오 등록(랭크) 승률, 소환사 아이콘 추가

hubaek opened the pull request:
## ✨ 담당 파트
- 롤 듀오 찾기 등록

## 🔎 작업 상세 내용
- wins, losses 계산해서 승률 반환
- profileIcon 필드에 String URL로 이미지 경로를 반환하게 함 

## 🔧 앞으로의 과제
- KDA 계산예정

## ✅ 테스트 코드 작성 및 기능 테스트 여부
- [] 테스트 코드 작성
- [x] 기능 테스트 여부

## ➕ 이슈 링크
- #21

hubaek committed:
fix(QueueType): RiotAPI 반환값에 따른 Enum 값 변경 #21

- SOLO, FLEX 일경우 넘겨 받는 값이 정해져 있어서 해당 값으로 로직 변경

hubaek committed:
fix(RiotClient): extractLeagueInfo 타입 변경 #21

- LeagueEntryResponse : List타입으로 반환되게 변경,

hubaek committed:
fix(RiotClientTest): Test3 타입 변경 #21

- LeagueEntryResponse : List타입으로 반환되게 변경

hubaek committed:
fix(Duo): targetRole 타입 변경 #21

- List<Lane> targetRoles -> Lane targetRole으로 리스트타입에서 단건으로 변경

hubaek committed:
feat(DuoController): duo 등록 추가 #21

- duo 등록 API 추가

hubaek committed:
feat(DuoService): duo 랭크 등록 추가 #21

- 큐타입이 SOLO, FLEX 일때 듀오 등록하는 기능 추가
- 추후 빠른 대전 등록 기능 추가 예정

hubaek committed:
feat(AccountRepository): memberId 기준 쿼리메서드 추가 #21

- memberId로 Account 객체를 찾는 쿼리메소드 추가

hubaek committed:
fix(JwtFilter): 마지막 로직에서 doFilter 누락 추가 #21

hubaek committed:
feat(Duo): 승률계산 로직 추가 #21

- Duo 엔티티의 승과 패로 승률을 계산을 하므로 Duo엔티티에 승률을 가져오는 메서드 추가

hubaek committed:
refactor(DuoService): LeagueEntryResponse 메서드 화 #21

- 요청 QueueType에 따라 가져오는 LeagueEntryResponse를 중복코드를 메서드화

hubaek committed:
feat(DuoService): 승률 반환 로직 추가 #21

- Duo 엔티티의 승률을 계산해서 Response에 반환

hubaek committed:
feat(DuoCreateResponse): response에 winRate 추가 #21

- winRate는 통계 컬럼으로 Response에만 컬럼으로 추가해서 반환

hubaek committed:
Refactor(Duo): profileIcon 타입 변경 #21

- Long -> String
- 프로필아이콘의 URL로 이미지를 반환을 해줌

hubaek committed:
feat(VersionRepository): 최신버전 찾는 쿼리메소드 추가 #21

hubaek committed:
feat(DuoService): profileIcon 반환 추가 #21

- profileIconUrl 값을 가져와서 아이콘의 이미지Url을 DuoResponse에 담아서 반환

hubaek committed:
feat(DuoCreateResponse): profileIcon 필드 추가 #21

- String profileIcon : profileIconUrl 값을 담아서 반환할 필드 추가

yokxim2 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/entity/Duo.java:
```diff
@@ -188,16 +177,23 @@ public static Duo flexOf(QueueType queueType,
+        }
+        return ((wins * 100) / (wins + losses));
+    }
 }
```
승률 계산을 소수점 셋째자리에서 반올림해서 사용하는 게 좋을 거 같습니다!
아까 점심 때 은영님하고도 얘기 잠깐 나눴는데, 이렇게 바꿔 주시면 될 것 같아요.

```
public double calculateWinRate(int wins, int losses) {
    if (wins + losses == 0) {
        return 0.0;
    }
    double winRate = (double) wins / (wins + losses) * 100;
    return BigDecimal.valueOf(winRate)
                     .setScale(3, RoundingMode.HALF_UP)
                     .doubleValue();
}
```

yokxim2 reviewed (commented):
메서드 분리 관련해서 튜터님 질문 후 다시 공유주시면 좋을것 같습니다! 고생하셨어요.

hubaek committed:
refactor(DuoService): 메서드명 변경 가독성 향상 #21

- getRankInfo 메서드는 굳이 메서드 화 필요가 없다고 생각되어 없앰
- getSelectedRankInfo -> getSelectedRankInfoByQueueType 메서드 명 변경

yokxim2 reviewed (approved):
좋습니다!

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/entity/Duo.java:
```diff
@@ -188,16 +177,23 @@ public static Duo flexOf(QueueType queueType,
         this.mic = mic;
     }
+
+    public int calculateWinRate(int wins, int losses) {
+        if (wins + losses == 0) {
+            return 0;
+        }
+        return ((wins * 100) / (wins + losses));
+    }
```
승률을 %로 가져가나요?
%로 가져간다면, 해당 로직이 맞는 것 같습니다.

해당 부분은 어떻게 진행하시는지 리뷰 보시고 답변 부탁드립니다.

hubaek replied:
승률이기 때문에 50% 같이 %로 가져갑니다

hubaek committed:
refactor(DuoService): puuid 가져오는 방식 변경 #21

- AccountDetail에 있는 puuid를 가져오는 방법으로 변경

hubaek committed:
Merge branch 'dev' into feat/duo/create-21

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/service/DuoService.java:
```diff
@@ -1,7 +1,128 @@
+        }
+        duoRepository.save(duo);
+        int winRate = duo.calculateWinRate(selectedRankInfo.getWins(), selectedRankInfo.getLosses());
+        return new DuoCreateResponse(duo, winRate);
```
승률을 ResponseDto에 담아 주셨네요.
CreateResponseDto 를 보니 승률만 있고 플레이한 게임 횟수가 없던데, 승리한 게임 횟수는 어떻게 반환 받을 수 있나요?

승률에 대한 계산보다는 **승리한 횟수, 패배한 횟수, 전체 플레이한 게임 횟수**를 response 해 주는 것이 나을 것 같습니다.

hubaek replied:
Duo엔티티 필드에 있는 승리한 횟수, 패배한 횟수 반환, 승률 계산해서 response 반환 하겠습니다.
전체 플레이한 횟수는 의미가 없어서 넣을 필요는 없습니다.

hubaek committed:
feat(DuoCreateResponse): 누락된 필드 추가 #21

- 랭크와 관련된 필드 : tier, ranks 추가
- 승,패 필드 : wins, losses 추가

for2ver0 reviewed (changes requested):
해당 리뷰에 대한 답변 부탁드립니다.

hubaek committed:
fix(DuoService): getLatestVersion 검증 로직 변경 #21

- 의미없는 예외처리 변경

hubaek committed:
fix(DuoService): getProfileIconUrl 메서드 오류 변경  #21

- puuid 값을 받아와야하는데 summonerId값을 받아와서 오류 해결
